src/algebra/category/FinVect.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2021 Jakob von Raumer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob von Raumer
+-/
+import category_theory.monoidal.rigid
+import linear_algebra.tensor_product_basis
+import linear_algebra.coevaluation
+import .Module.monoidal
+
+/-!
+# The category of finite dimensional vector spaces
+
+This introduces `FinVect K`, the category of finite dimensional vector spaces on a field `K`.
+It is implemented as a full subcategory on a subtype of  `Module K`.
+We first create the instance as a category, then as a monoidal category and then as a rigid monoidal
+category.
+
+## Future work
+
+* Show that `FinVect K` is a symmetric monoidal category.
+
+-/
+noncomputable theory
+
+open category_theory Module.monoidal_category
+open_locale classical big_operators
+
+universes u
+
+variables (K : Type u) [field K]
+
+/-- Define `FinVect` as the subtype of `Module.{u} K` of finite dimensional vector spaces. -/
+@[derive [category, has_coe_to_sort]]
+def FinVect := { V : Module.{u} K // finite_dimensional K V }
+
+namespace FinVect
+
+instance finite_dimensional (V : FinVect K): finite_dimensional K V := V.prop
+
+instance : inhabited (FinVect K) := ⟨⟨Module.of K K, finite_dimensional.finite_dimensional_self K⟩⟩
+
+protected def coe_comp {U V W : FinVect K} (f : U ⟶ V) (g : V ⟶ W) :
+  ((f ≫ g) : U → W) = (g : V → W) ∘ (f : U → V) := rfl
+
+instance monoidal_category : monoidal_category (FinVect K) :=
+monoidal_category.full_monoidal_subcategory
+  (λ V, finite_dimensional K V)
+  (finite_dimensional.finite_dimensional_self K)
+  (λ X Y hX hY, by exactI finite_dimensional_tensor_product X Y)
+
+variables (V : FinVect K)
+
+/-- The dual module is the dual in the rigid monoidal category `FinVect K`. -/
+def FinVect_dual : FinVect K :=
+⟨Module.of K (module.dual K V), subspace.module.dual.finite_dimensional⟩
+
+instance : has_coe_to_fun (FinVect_dual K V) :=
+{ F := λ v, V → K,
+  coe := λ v, by { change V →ₗ[K] K at v, exact v, }, }
+
+open category_theory.monoidal_category
+
+/-- The coevaluation map is defined in `linear_algebra.coevaluation`. -/
+def FinVect_coevaluation : 𝟙_ (FinVect K) ⟶ V ⊗ (FinVect_dual K V) :=
+by { haveI := V.prop, exact coevaluation K V }
```

Comment:
I don't really get why this needs the `haveI`.

Reply:
The problem is that `coevaluation` needs a `finite_dimensional` typeclass available. The `haveI` first notes a new hypothesis, and then resets the instance cache (so that typeclass inference can "see" that new hypothesis).

(Happily this issue is apparently going away in Lean4.)

Reply:
Oh, I see, but your `instance finite_dimensional ...` should be taking care of this. :-(

Reply:
The problem is that we have multiple coercions here. We have the `has_coe_to_sort`, but we also have the coercion from a subtype, giving us two route to `Type`: via the `@[derive has_coe_to_sort]`, but also via `Module K`, and *its* `has_coe_to_sort`.

Reply:
I have to run, but could you investigate if `instance foo : has_coe (FinVect.{u} K) (Module.{u} K) := { coe := λ V, V.1, }
` helps?

Reply:
(It does work: add that instance, then you can just write `by apply coevaluation K V` here.

Reply:
Oh, so in principle that coercion from `FinVect` to `Module` exists, but the definition of `FinVect` isn't transparent enough for Lean to get it.